### PR TITLE
Add `unique` function.

### PR DESCRIPTION
**Pull request by matthiasdiener**
*Please squash*

### CODE REVIEW

**Comment by inducer** [comment]
https://docs.python.org/3/library/itertools.html#itertools.groupby?

**Comment by matthiasdiener** [comment]
> https://docs.python.org/3/library/itertools.html#itertools.groupby?

As far as I can see, `groupby` only removes *consecutive* duplicates, while `unique` removes all duplicates:

```python
>>> from itertools import groupby
>>> [k for k, g in groupby('AAAABBBCCDAABBB')]
['A', 'B', 'C', 'D', 'A', 'B']

>>> from pytools import unique
>>> list(unique('AAAABBBCCDAABBB'))
['A', 'B', 'C', 'D']
```

File: pytools/__init__.py [hunk 1]
```diff
@@ -2994,6 +2999,15 @@ def strtobool(val: Optional[str], default: Optional[bool] = None) -> bool:
 # }}}
 
 
+# {{{ unique
+
+def unique(seq: Sequence[T]) -> Iterator[T]:
+    """Yield unique elements in *seq*. See also :func:`reversed`."""
+    yield from list(dict.fromkeys(seq))
```
Comment (inducer) [inline pytools/__init__.py]:
```suggestion
    return iter(dict.fromkeys())
```

Comment (inducer) [reply pytools/__init__.py]: Saw your simplification: `yield from` is not zero cost, `return` is.

**Comment by inducer** [comment]
> As far as I can see, `groupby` only removes _consecutive_ duplicates, while `unique` removes all duplicates:

That's a fair point. Maybe emphasize that in the docstring?

File: pytools/__init__.py (hunk 1 above)
Comment (matthiasdiener) [reply pytools/__init__.py]: > Saw your simplification: `yield from` is not zero cost, `return` is.

Thanks, `return iter` is indeed much faster. Done in [d4f9ab3](https://github.com/inducer/pytools/pull/187/commits/d4f9ab3135ddde4ccc1810041b9770865fdb6ade)

**Comment by matthiasdiener** [comment]
> > As far as I can see, `groupby` only removes _consecutive_ duplicates, while `unique` removes all duplicates:
> 
> That's a fair point. Maybe emphasize that in the docstring?

What do you think of [540e565](https://github.com/inducer/pytools/pull/187/commits/540e565fe6e4e11055b529a469b6e102afb10479)?

**Comment by inducer** [comment]
LGTM, thanks! Do you need that released?

**Pull request merged**